Cache MethodToCallFromOutside and create blueprints in parallel

srmurguia opened the pull request:
The function getMethodToCallForDependency is called on every dependency.
This function creates a new ModuleBluePrint to extract
methodToCallFromOutside. In a project with N modules there can be upto
N*(N-1)/2 dependencies, meaning that the same number of ModuleBluePrint will
be generated and deleted.

This change adds a cache to store the values of methodToCallFromOutside,
needing now only N ModuleBluePrint.

Also, now the constructor of ProjectBlueprint create moduleBlueprints in
parallel.

A project with 1000 modules went from taking around 100 seconds to just
18 (5x imrpovement in time).

srmurguia committed:
Cache MethodToCallFromOutside and create blueprints in parallel

The function getMethodToCallForDependency is called on every dependency.
This function creates a new ModuleBluePrint to extract
methodToCallFromOutside. In a project with N modules there can be upto
N*(N-1)/2 dependencies, meaning that the same number of ModuleBluePrint will
be generated and deleted.

This change adds a cache to store the values of methodToCallFromOutside,
needing now only N ModuleBluePrint.

Also, now the constructor of ProjectBlueprint create moduleBlueprints in
parallel.

A project with 1000 modules wnet from taking around 100 seconds to just
18 (5x imrpovement in time).

srmurguia committed:
Cache MethodToCallFromOutside and create blueprints in parallel

The function getMethodToCallForDependency is called on every dependency.
This function creates a new ModuleBluePrint to extract
methodToCallFromOutside. In a project with N modules there can be upto
N*(N-1)/2 dependencies, meaning that the same number of ModuleBluePrint will
be generated and deleted.

This change adds a cache to store the values of methodToCallFromOutside,
needing now only N ModuleBluePrint.

Also, now the constructor of ProjectBlueprint create moduleBlueprints in
parallel.

A project with 1000 modules wnet from taking around 100 seconds to just
18 (5x imrpovement in time).

borisf committed:
init drop

borisf committed:
added fixes

NikitaKozlov committed:
Merge pull request #48 from borisf/travis-integ

Travis integ

borisf committed:
added third party with licenses

borisf committed:
verified licenses for all src files

borisf committed:
added contributing doc

borisf committed:
Merge pull request #50 from borisf/release

Release

NikitaKozlov commented on src/main/kotlin/com/google/androidstudiopoet/ModuleBlueprintFactory.kt:
```diff
@@ -24,7 +28,15 @@ object ModuleBlueprintFactory {
             WARNING: creation of ModuleBlueprint could be expensive
          */
-        return ModuleBlueprint(index, getModuleNameByIndex(index), projectRoot, listOf(), listOf(), config).methodToCallFromOutside
+        synchronized(methodLock[index]) {
```
Can you please extract this whole block into a separate method? You can call it something like `getMethodToCallForModule`, then we'll be able to optimize it even more without touching the main method.

srmurguia replied:
Thanks for this suggestion, it makes a lot of sense to have it in a function. Hopefully this function will be faster than creating and throwing a ModuleBlueprint once we implement it :p

NikitaKozlov replied:
In the long run, I hope we can build modules in the order that doesn't require to have a cache at all unless there is a circular dependency. :)

srmurguia committed:
Merge branch 'usemap' of https://github.com/srmurguia/java-generator into usemap

srmurguia committed:
Create getMethodToCallForModule as suggested by Nikita

Right now this method simply generates a new instance of
ModuleBlueprint to only extract methodToCallFromOutside. This function
could be improve to generate the same result in a faster way without
needing to change other parts of the code.
